src/main/java/org/squiddev/cobalt/lib/StringLib.java

```diff
@@ -603,6 +611,392 @@ static Varargs match(LuaState state, Varargs args) throws LuaError {
 		return str_find_aux(state, args, false);
 	}
 
+	private static int packint(long num, int size, byte[] output, int offset, int alignment, ByteOrder endianness, boolean signed) {
+		int total_size = 0;
+		if (offset % Math.min(size, alignment) != 0 && alignment > 1) {
+			for (int i = 0; offset % Math.min(size, alignment) != 0 && i < alignment; i++) {
+				output[offset++] = 0;
+				total_size++;
+			}
+		}
+		if (endianness == ByteOrder.BIG_ENDIAN) {
+			int added_padding = 0;
+			if (size > 8) for (int i = 0; i < size - 8; i++) {
+				output[offset + i] = signed && (num & (1 << (size * 8 - 1))) != 0 ? (byte)0xFF : 0;
+				added_padding++;
+				total_size++;
+			}
+			for (int i = added_padding; i < size; i++) {
+				output[offset + i] = (byte) ((num >> ((size - i - 1) * 8)) & 0xFF);
+				total_size++;
+			}
+		} else {
+			for (int i = 0; i < Math.min(size, 8); i++) {
+				output[offset + i] = (byte) ((num >> (i * 8)) & 0xFF);
+				total_size++;
+			}
+			for (int i = 8; i < size; i++) {
+				output[offset + i] = signed && (num & (1 << (size * 8 - 1))) != 0 ? (byte)0xFF : 0;
+				total_size++;
+			}
+		}
+		return total_size;
+	}
+
+	private static int packoptsize(byte opt, int alignment) {
+		int retval = 0;
+		switch (opt) {
+			case 'b':
+			case 'B':
+			case 'x':
+				retval = 1;
+				break;
+			case 'h':
+			case 'H':
+				retval = 2;
+				break;
+			case 'f':
+			case 'j': // lua_Integer is 32-bit because of bit32 support
+			case 'J':
+				retval = 4;
+				break;
+			case 'l':
+			case 'L':
+			case 'T':
+			case 'd':
+			case 'n':
+				retval = 8;
+				break;
+		}
+		if (alignment > 1 && retval % alignment != 0) retval += (alignment - (retval % alignment));
+		return retval;
+	}
+
+	/**
+	 * string.pack (fmt, v1, v2, ...)
+	 *
+	 * Returns a binary string containing the values v1, v2, etc.
+	 * serialized in binary form (packed) according to the format string fmt.
+	 */
+	static Varargs pack(Varargs args) throws LuaError {
+		LuaString fmt = args.arg(1).checkLuaString();
+		ByteOrder endianness = ByteOrder.nativeOrder();
+		int alignment = 1;
+		int currentSize = 64, pos = 0;
+		int argnum = 2;
+		byte[] output = new byte[currentSize];
```

Comment:
I wonder if it's worth pre-computing the size (using `packsize` function), and then using a `ByteBuffer` here? This way we don't need to worry about resizing, and _think_ can just use `putInt` instead of rolling our own.

Reply:
I decided to have it resize the buffer instead of using `packsize` to pre-allocate the required size to avoid looping over the format string twice.

Reply:
Yeah, I understand, I guess one could use Cobalt's `Buffer` instead - gives the same advantages, just no `putFloat`, etc... methods.

One thing I was debating was whether it was worth maintaining a small cache (i.e. 16 elements) for string → pack sizes. Kinda want to do the same for format strings at some point.